File: apps/web/src/types/watchlist.ts
// Watchlist & Alert domain types

export type WatchlistType =
  | 'email'
  | 'phone'
  | 'domain'
  | 'url'
  | 'ip_address'
  | 'sha256'
  | 'md5'
  | 'crypto_wallet'
  | 'bank_account'
  | 'vehicle_registration'
  | 'passport'
  | 'device_id'
  | 'imei'
  | 'mac_address'
  | 'regex'
  | 'keyword'

export type AlertType =
  | 'exact_match'
  | 'regex_match'
  | 'fuzzy_match'
  | 'cross_case_match'
  | 'high_risk_match'
  | 'repeated_appearance'
  | 'ai_alert'
  | 'manual_alert'
  | 'system_alert'

export type AlertSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info'

export type AlertStatus = 'new' | 'acknowledged' | 'resolved' | 'dismissed'

export type NotificationLevel = 'info' | 'warning' | 'error' | 'critical'

// ── Watchlist ──────────────────────────────────────────────────────────────────

export interface WatchlistRead {
  id: string
  name: string
  description: string | null
  watchlistType: WatchlistType
  isActive: boolean
  caseId: string | null
  createdById: string | null
  createdByEmail: string | null
  entryCount: number
  alertCount: number
  createdAt: string
  updatedAt: string
}

export interface WatchlistListResponse {
  items: WatchlistRead[]
  total: number
  page: number
  pages: number
}

export interface WatchlistCreate {
  name: string
  description?: string | null
  watchlistType: WatchlistType
  isActive?: boolean
  caseId?: string | null
}

export interface WatchlistUpdate {
  name?: string
  description?: string | null
  watchlistType?: WatchlistType
  isActive?: boolean
}

export interface WatchlistStats {
  totalWatchlists: number
  activeWatchlists: number
  totalEntries: number
  totalAlerts: number
  alertsToday: number
  alertsThisWeek: number
  topHitWatchlists: Array<{ id: string; name: string; alertCount: number }>
}

// ── WatchlistEntry ─────────────────────────────────────────────────────────────

export interface WatchlistEntryRead {
  id: string
  watchlistId: string
  value: string
  normalizedValue: string
  isRegex: boolean
  description: string | null
  isActive: boolean
  hitCount: number
  createdById: string | null
  createdAt: string
  updatedAt: string
}

export interface WatchlistEntryListResponse {
  items: WatchlistEntryRead[]
  total: number
}

export interface WatchlistEntryCreate {
  value: string
  isRegex?: boolean
  description?: string | null
}

// ── WatchlistAlert ─────────────────────────────────────────────────────────────

export interface WatchlistAlertRead {
  id: string
  watchlistId: string | null
  watchlistEntryId: string | null
  evidenceId: string | null
  caseId: string
  alertType: AlertType
  severity: AlertSeverity
  title: string
  description: string | null
  matchedValue: string | null
  matchedEntityType: string | null
  confidence: number
  status: AlertStatus
  isCrossCase: boolean
  crossCaseCount: number
  crossCaseAccessible: boolean
  alertMetadata: Record<string, unknown>
  acknowledgedAt: string | null
  resolvedAt: string | null
  createdAt: string
  updatedAt: string
  watchlistName: string | null
  evidenceFilename: string | null
  caseReference: string | null
}

export interface WatchlistAlertListResponse {
  items: WatchlistAlertRead[]
  total: number
  page: number
  pages: number
  newCount: number
  criticalCount: number
}

export interface AlertStats {
  total: number
  newCount: number
  acknowledgedCount: number
  resolvedCount: number
  dismissedCount: number
  criticalCount: number
  highCount: number
  crossCaseCount: number
  alertsToday: number
  alertsThisWeek: number
}

// ── AlertNotification ──────────────────────────────────────────────────────────

export interface AlertNotificationRead {
  id: string
  alertId: string | null
  caseId: string | null
  title: string
  message: string | null
  level: NotificationLevel
  isRead: boolean
  isArchived: boolean
  readAt: string | null
  createdAt: string
}

export interface NotificationListResponse {
  items: AlertNotificationRead[]
  total: number
  unreadCount: number
}

export interface NotificationCount {
  unreadCount: number
}

File: apps/api/app/schemas/watchlist.py
"""Watchlist, WatchlistEntry, WatchlistAlert, and AlertNotification schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.base import BaseSchema


# ── Watchlist ─────────────────────────────────────────────────────────────────

class WatchlistCreate(BaseSchema):
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    watchlist_type: str
    is_active: bool = True
    case_id: uuid.UUID | None = None


class WatchlistUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    watchlist_type: str | None = None
    is_active: bool | None = None


class WatchlistRead(BaseSchema):
    id: uuid.UUID
    name: str
    description: str | None
    watchlist_type: str
    is_active: bool
    case_id: uuid.UUID | None
    created_by_id: uuid.UUID | None
    created_by_email: str | None = None
    entry_count: int = 0
    alert_count: int = 0
    created_at: datetime
    updated_at: datetime


class WatchlistListResponse(BaseSchema):
    items: list[WatchlistRead]
    total: int
    page: int
    pages: int


# ── WatchlistEntry ────────────────────────────────────────────────────────────

class WatchlistEntryCreate(BaseSchema):
    value: str = Field(..., min_length=1)
    is_regex: bool = False
    description: str | None = None


class WatchlistEntryUpdate(BaseSchema):
    value: str | None = Field(default=None, min_length=1)
    is_regex: bool | None = None
    description: str | None = None
    is_active: bool | None = None


class WatchlistEntryRead(BaseSchema):
    id: uuid.UUID
    watchlist_id: uuid.UUID
    value: str
    normalized_value: str
    is_regex: bool
    description: str | None
    is_active: bool
    hit_count: int
    created_by_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime


class WatchlistEntryListResponse(BaseSchema):
    items: list[WatchlistEntryRead]
    total: int


# ── WatchlistAlert ────────────────────────────────────────────────────────────

class WatchlistAlertRead(BaseSchema):
    id: uuid.UUID
    watchlist_id: uuid.UUID | None
    watchlist_entry_id: uuid.UUID | None
    evidence_id: uuid.UUID | None
    case_id: uuid.UUID
    alert_type: str
    severity: str
    title: str
    description: str | None
    matched_value: str | None
    matched_entity_type: str | None
    confidence: float
    status: str
    is_cross_case: bool
    # cross_case_ids contains str UUIDs; RBAC-filtered message added in API layer
    cross_case_count: int = 0
    cross_case_accessible: bool = False
    alert_metadata: dict[str, Any] = {}
    acknowledged_at: datetime | None
    resolved_at: datetime | None
    created_at: datetime
    updated_at: datetime
    # Denormalized display fields
    watchlist_name: str | None = None
    evidence_filename: str | None = None
    case_reference: str | None = None


class WatchlistAlertListResponse(BaseSchema):
    items: list[WatchlistAlertRead]
    total: int
    page: int
    pages: int
    new_count: int = 0
    critical_count: int = 0


class AlertStats(BaseSchema):
    total: int
    new_count: int
    acknowledged_count: int
    resolved_count: int
    dismissed_count: int
    critical_count: int
    high_count: int
    cross_case_count: int
    alerts_today: int
    alerts_this_week: int


# ── AlertNotification ─────────────────────────────────────────────────────────

class AlertNotificationRead(BaseSchema):
    id: uuid.UUID
    alert_id: uuid.UUID | None
    case_id: uuid.UUID | None
    title: str
    message: str | None
    level: str
    is_read: bool
    is_archived: bool
    read_at: datetime | None
    created_at: datetime


class NotificationListResponse(BaseSchema):
    items: list[AlertNotificationRead]
    total: int
    unread_count: int


class NotificationCount(BaseSchema):
    unread_count: int


# ── Watchlist statistics ──────────────────────────────────────────────────────

class WatchlistStats(BaseSchema):
    total_watchlists: int
    active_watchlists: int
    total_entries: int
    total_alerts: int
    alerts_today: int
    alerts_this_week: int
    top_hit_watchlists: list[dict[str, Any]] = []
